### PR TITLE
Support flying harvesters

### PR DESCRIPTION
**Pull request by PunkPun**
Adds some of the plumbing needed for docking logic to support aircraft #20168. With this PR openra will support chinook harvesters from Generals

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/DeliverResources.cs [hunk 1]
```diff
@@ -30,6 +32,7 @@ public DeliverResources(Actor self, Actor targetActor = null)
 		{
 			movement = self.Trait<IMove>();
 			harv = self.Trait<Harvester>();
+			aircraft = self.TraitOrDefault<Aircraft>();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/DeliverResources.cs]:
```suggestion
			aircraft = movement as Aircraft;
```

File: OpenRA.Mods.Common/Activities/FindAndDeliverResources.cs [hunk 2]
```diff
@@ -38,7 +38,8 @@ public FindAndDeliverResources(Actor self, Actor deliverActor = null)
 		{
 			harv = self.Trait<Harvester>();
 			harvInfo = self.Info.TraitInfo<HarvesterInfo>();
-			mobile = self.Trait<Mobile>();
+			mobile = self.TraitOrDefault<Mobile>();
+			move = self.Trait<IMove>();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/FindAndDeliverResources.cs]:
```suggestion
			move = self.Trait<IMove>();
			mobile = move as Mobile;
```

File: OpenRA.Mods.Common/Activities/HarvestResource.cs [hunk 3]
```diff
@@ -32,6 +33,7 @@ public class HarvestResource : Activity
 		public HarvestResource(Actor self, CPos targetCell)
 		{
 			harv = self.Trait<Harvester>();
+			aircraft = self.TraitOrDefault<Aircraft>();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/HarvestResource.cs]:
```suggestion
			aircraft =move as Aircraft;
```
(And need to re-order this below the definition of `move`.)

File: OpenRA.Mods.Common/Traits/Buildings/Refinery.cs [hunk 4]
```diff
@@ -29,6 +29,9 @@ public class RefineryInfo : TraitInfo, Requires<WithSpriteBodyInfo>, IAcceptReso
 		[Desc("Docking cell relative to top-left cell.")]
 		public readonly CVec DockOffset = CVec.Zero;
 
+		[Desc("If harvester in an aircraft does it need to land to unload resources.")]
+		public readonly bool RequireAircratToLand = true;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Buildings/Refinery.cs]:
```suggestion
		public readonly bool RequireAircraftToLand = true;
```

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 5]
```diff
@@ -73,6 +73,9 @@ public class HarvesterInfo : ConditionalTraitInfo, Requires<MobileInfo>
 		[Desc("Does the unit queue harvesting runs instead of individual harvest actions?")]
 		public readonly bool QueueFullLoad = false;
 
+		[Desc("If harvester in an aircraft does it need to land to collect resources.")]
+		public readonly bool RequireAircratToLand = true;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Harvester.cs]:
```suggestion
		public readonly bool RequireAircraftToLand = true;
```

**Comment by PunkPun** [comment]
I'll fix it up tomorrow when i get home

File: OpenRA.Mods.Common/Activities/DeliverResources.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Activities/DeliverResources.cs]: I've used `move is Aircraft` instead

File: OpenRA.Mods.Common/Traits/Buildings/Refinery.cs (hunk 4 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Buildings/Refinery.cs]: typo fixed

**Comment by AspectInteractive** [comment]
Found a typo in FindAndDeliverResources.cs, `FindTileInCicle` should be `FindTileInCircle`.

File: OpenRA.Mods.Common/Traits/Buildings/Refinery.cs [hunk 6]
```diff
@@ -75,6 +78,7 @@ public class Refinery : INotifyCreated, ITick, IAcceptResources, INotifySold, IN
 		public bool IsDragRequired => info.IsDragRequired;
 		public WVec DragOffset => info.DragOffset;
 		public int DragLength => info.DragLength;
+		public bool RequireAircratToLand => info.RequireAircraftToLand;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Buildings/Refinery.cs]:
```suggestion
		public bool RequireAircraftToLand => info.RequireAircraftToLand;
```

File: OpenRA.Mods.Common/Traits/Buildings/Refinery.cs [hunk 7]
```diff
@@ -29,6 +29,9 @@ public class RefineryInfo : TraitInfo, Requires<WithSpriteBodyInfo>, IAcceptReso
 		[Desc("Docking cell relative to top-left cell.")]
 		public readonly CVec DockOffset = CVec.Zero;
 
+		[Desc("If harvester in an aircraft does it need to land to unload resources.")]
+		public readonly bool RequireAircraftToLand = true;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:33]: This is not even used here and odd for the default mods. Maybe use an existing trait from `Aircraft` or put this onto an `AircraftHarvester`.

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Buildings/Refinery.cs:33]: We want this to be a choice on a per dock basis, so it needs to stay on `Refinery` until `Dock` class is added

**Comment by PunkPun** [comment]
I'll address #16163 here

**Comment by PunkPun** [comment]
Closing this PR for now. I have noticed a bunch of issues deeper in the engine that make this PR messier than it needs to be

I will reopen once I have solved those issues or we are closer to finishing the refactor

**Pull request closed without merging**